include/h2o/httpclient.h

```diff
@@ -0,0 +1,109 @@
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+#ifndef h2o__httpclient_h
+#define h2o__httpclient_h
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "h2o/header.h"
+#include "h2o/socket.h"
+#include "h2o/socketpool.h"
+#include "h2o/timeout.h"
+
+typedef struct st_h2o_httpclient_t h2o_httpclient_t;
+
+typedef struct st_h2o_httpclient_features_t {
+    h2o_iovec_t *proxy_protocol;
+    int *chunked;
+    int connection_header;
+} h2o_httpclient_features_t;
+
+typedef void (*h2o_httpclient_proceed_req_cb)(h2o_httpclient_t *client, size_t written, int is_end_stream);
+typedef int (*h2o_httpclient_body_cb)(h2o_httpclient_t *client, const char *errstr);
+typedef h2o_httpclient_body_cb (*h2o_httpclient_head_cb)(h2o_httpclient_t *client, const char *errstr, int minor_version,
+                                                           int status, h2o_iovec_t msg, h2o_header_t *headers,
+                                                           size_t num_headers, int rlen);
+typedef h2o_httpclient_head_cb (*h2o_httpclient_connect_cb)(h2o_httpclient_t *client, const char *errstr, h2o_iovec_t *method, h2o_url_t *url,
+                                                              const h2o_header_t **headers, size_t *num_headers, h2o_iovec_t *body,
+                                                              h2o_httpclient_proceed_req_cb *proceed_req_cb, h2o_httpclient_features_t features, h2o_url_t *origin);
+typedef int (*h2o_httpclient_informational_cb)(h2o_httpclient_t *client, int minor_version, int status, h2o_iovec_t msg,
+                                                h2o_header_t *headers, size_t num_headers);
+
+typedef struct st_h2o_httpclient_connection_pool_t {
+    /**
+     * used to establish connections and pool those when h1 is used
+     */
+    h2o_socketpool_t *socketpool;
+
+    struct {
+        h2o_linklist_t conns;
+    } http2;
+
+} h2o_httpclient_connection_pool_t;
+
+typedef struct st_h2o_httpclient_ctx_t {
```

Comment:
I would also appreciate it if you could clarify the role of the object as well as why it is a separate structure from `h2o_httpclient_connection_pool_t`.

My understanding is that they are separate because this ctx_t can be for example path-specific (having different io-timeout for every path). Am I correct?